forem/forem
# Message context menu in connect is not accessible #12250

## What type of PR is this? (check all applicable)


## Description

In Connect, the message options menu (the three dots at the right) previously appeared only on hover, and was inaccessible for keyboard and assistive technology users. This PR makes sure those buttons are 'real' buttons which become visible on focus as well as hover.

Other accessibility features have been added also, namely:

- The buttons are identified as having a popup with an expanded state
- The menus can be opened on keyboard Enter press as well as click
- The menus can be dismissed with the Escape key
- When opened, focus is transferred directly to the first item in the list
- When closed, focus is transferred back to the button which opened the menu
- Where the dropdown had two items, they've been changed to an unordered list so a screen reader announces how many actions are available

I've also made sure the menu closes if a user Tabs away from the menu options, in which case the trigger button is _not_ auto-focused (since the user is moving away from the menu options, we don't want to steal focus back)

### Some additional context

Dropdowns as a whole are going to be looked at in #12297. I've tried to avoid any hasty abstractions here, so when we pick up that issue we can look at the bigger picture.

## Related Tickets & Documents

Closes #12250

## QA Instructions, Screenshots, Recordings

- Start a chat in `/connect` (you may need to set up your local instance to have some mutual follows to do so)
- Write a message and send to the chat
- I found I needed to reload the page to see my sent message (not sure at the moment if that's just a problem for me locally, but mentioning just in case!)
- Check that if you hover with a mouse towards the right hand side, you can see the 'three dots' menu appear next to your message
- Check that you can click the menu for it to open
- Check that you can close the menu with Escape key
- Check that you can Tab to the menu button, and it's visible when focused
- Press Enter to open the menu and check that the first item (Edit) is focused
- Check you can Tab to the second item (Delete)
- Check that pressing Escape closes the menu and refocuses the button with the three dots
- Check that if you re-open the menu with the keyboard, and tab _past_ the final option, the menu closes but doesn't place focus on the three dots button
- Check that when you focus on the three dots button with a screen reader on, that it's announced as "expanded" when the dropdown is open, and "collapsed" when it's not (exact behaviour may vary with different screen readers/browsers)
- Check that even when the three dots button isn't currently visible, that it's exposed to screen readers (e.g. in the rotor)
- Log out, and log in as the chat recipient, and check the same behaviours hold true for the "three dots" button on someone else's message (in which case 'Report abuse' is the only option in the popup)

https://user-images.githubusercontent.com/20773163/118838785-e083dd00-b8bd-11eb-96f3-0b64afd91c6b.mp4



### UI accessibility concerns?

Covered in the above test steps

## Added tests?

      have not been included_

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

N/A

## [optional] What gif best describes this PR or how it makes you feel?

![much better](https://media.giphy.com/media/ieh305yo7DKqC5Mrla/giphy.gif)
